src/inference/amh.jl

```diff
@@ -0,0 +1,278 @@
+                vi[vns[1]] = vals[1]
+            else
+                vi[vns[1]] = [vals[1]]
+            end
+        elseif !(v_isarr)
+            vi[vns[1]] = [vals]
+        else
+            error("Cannot assign `NamedTuple` to `VarInfo`")
+        end
+    end
+end
+
+"""
+    gen_logπ_mh(vi::VarInfo, spl::Sampler, model)   
+
+Generate a log density function -- this variant uses the 
+`set_namedtuple!` function to update the `VarInfo`.
+"""
+function gen_logπ_mh(spl::Sampler, model)
+    function logπ(x)::Float64
+        vi = spl.state.vi
+        x_old, lj_old = vi[spl], vi.logp
+        # vi[spl] = x
+        set_namedtuple!(vi, x)
+        runmodel!(model, vi)
+        lj = vi.logp
+        vi[spl] = x_old
+        setlogp!(vi, lj_old)
+        return lj
+    end
+    return logπ
+end
+
+"""
+    dist_val_tuple(spl::Sampler{<:MH})
+
+Returns two `NamedTuples`. The first `NamedTuple` has symbols as keys and distributions as values.
+The second `NamedTuple` has model symbols as keys and their stored values as values.
+"""
+function dist_val_tuple(spl::Sampler{<:MH})
+    vns = _getvns(spl.state.vi, spl)
+    dt = _dist_tuple(spl.state.vi.metadata, spl.state.vi, vns)
+    vt = _val_tuple(spl.state.vi.metadata, spl.state.vi, vns)
+    return dt, vt
+end
+
+@generated function _val_tuple(metadata::NamedTuple, vi::VarInfo, vns::NamedTuple{names}) where {names}
+    length(names) === 0 && return :(NamedTuple())
+    expr = Expr(:tuple)
+    map(names) do f
+        push!(expr.args, Expr(:(=), f, :(getindex.(Ref(vi), metadata.$f.vns))))
+    end
+    return expr
+end
+
+@generated function _dist_tuple(metadata::NamedTuple, vi::VarInfo, vns::NamedTuple{names}) where {names}
+    length(names) === 0 && return :(NamedTuple())
+    expr = Expr(:tuple)
+    map(names) do f
+        push!(expr.args, Expr(:(=), f, :(metadata.$f.dists[1])))
+    end
+    return expr
+end
+
+###############################
+# Static MH (from prior only) #
+###############################
+
+function Sampler(
+    alg::MH,
+    model::Model,
+    s::Selector=Selector()
+)
+    # Set up info dict.
+    info = Dict{Symbol, Any}()
+
+    # Make a varinfo.
+    vi = VarInfo(model)
+
+    # Make a density model.
+    dm = AMH.DensityModel(x -> 0.0)
+
+    # Set up state struct.
+    state = MHState(model, dm)
+
+    # Generate a sampler.
+    spl = Sampler(alg, info, s, state)
+
+    # Update the density model.
+    spl.state.density_model = AMH.DensityModel(gen_logπ_mh(spl, model))
+
+    return spl
+end
+
+function sample_init!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:MH},
+    N::Integer;
+    verbose::Bool=true,
+    resume_from=nothing,
+    kwargs...
+)
+    # Resume the sampler.
+    set_resume!(spl; resume_from=resume_from, kwargs...)
+
+    # Get `init_theta`
+    initialize_parameters!(spl; verbose=verbose, kwargs...)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:MH},
+    N::Integer;
```

Comment:
I think it might be better/safer to add a `::Nothing` argument as well, and implement the method that is further down in the call stack. Otherwise calling `step!(..., N; kwargs...)` and `step!(..., N, nothing; kwargs...)` return different results.